### PR TITLE
Replaced Vector with ArrayList, fixed Generics in Utils

### PR DESCRIPTION
**Pull request by miho**
**Question:** please let us know if you still maintain this codebase. If so, we are happy to send PRs. Here's a list of the most important fixes we addressed so far.

- Since this project does not use multithreading the usage of [Vector](https://docs.oracle.com/javase/8/docs/api/java/util/Vector.html) is not a good choice. [Vector](https://docs.oracle.com/javase/8/docs/api/java/util/Vector.html) is fully synchronized and introduces some overhead. That's why it has been replaced by [ArrayList](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html). 
- Some utility methods were missing generics `List` vs `List<String>` etc. This has been corrected.
- The current package names prevent APTED from being used as library since conflicts are very likely. There are too many `util` packages out there ;) We introduced package names derived from university & project name.
- gradle project has been updated: it was missing encoding and Java version which often causes trouble. We also specify the gradle wrapper version which is important to prevent compatibility issues.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,3 +14,4 @@ hs_err_pid*
 # Gradle stuff
 build/
 .gradle/
+/.nb-gradle/
```
Comment (mateuszpawlik) [inline .gitignore]: The file has not changed. This should not be part of the commit.

File: build.gradle [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
 repositories {
     mavenCentral()
 }
 
 // Project dependencies. Currently only for unit tests.
 dependencies {
-    testCompile 'junit:junit:4.12',
-                'com.google.code.gson:gson:2.8.0'
+    testCompile 'junit:junit:4.12';
+    testCompile 'com.google.code.gson:gson:2.8.0'
```
Comment (mateuszpawlik) [inline build.gradle]: What's the difference here? If only syntax, then it should go away.

File: build.gradle [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
 }
 
 // Specify main class for the JAR file.
 jar {
     manifest {
-        attributes 'Main-Class': 'util.CommandLine'
+        attributes 'Main-Class': 'at.unisalzburg.apted.util.CommandLine'
```
Comment (mateuszpawlik) [inline build.gradle]: I post it here, because it's the first occurrence of that change.
 
Since it's the second pull request about packaging, I've decided to go for it. There are however two issues.
(1) Will such change affect the current users of the library? If yes, what are the options to deal with it?
(2) I'd prefer 'at.unisalzburg.dbresearch' instead of 'at.unisalzburg' prefix.

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (mateuszpawlik) [inline gradle/wrapper/gradle-wrapper.properties]: This file is automatically generated and should go away from the repository.

File: gradlew [hunk 5]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Comment (mateuszpawlik) [inline gradlew]: This file is automatically generated and should go away from the repository.

File: gradlew.bat [hunk 6]
```diff
@@ -0,0 +1,84 @@
+@if "%DEBUG%" == "" @echo off
```
Comment (mateuszpawlik) [inline gradlew.bat]: This file is automatically generated and should go away from the repository.

File: settings.gradle [hunk 7]
```diff
@@ -0,0 +1 @@
+rootProject.name = 'apted'
```
Comment (mateuszpawlik) [inline settings.gradle]: The file has not changed. This should not be part of the commit.

File: src/main/java/at/unisalzburg/apted/distance/AllPossibleMappingsTED.java [hunk 8]
```diff
@@ -21,36 +21,36 @@
...
+import at.unisalzburg.apted.costmodel.CostModel;
 
 /**
  * Implements an exponential algorithm for the tree edit distance. It computes
  * all possible TED mappings between two trees and calculated their minimal
  * cost.
  *
- * @param C type of cost model.
- * @param D type of node data.
+ * @param <C></C> type of cost model.
```
Comment (mateuszpawlik) [inline src/main/java/at/unisalzburg/apted/distance/AllPossibleMappingsTED.java]: This change causes params to disappear from the generated javadoc. I think it should be <C> and <D> without the closing tags.

File: src/main/java/at/unisalzburg/apted/node/Node.java [hunk 9]
```diff
@@ -37,24 +38,24 @@
...
   /**
-   * Constructs a new node with the passed node data and an empty vector of
+   * Constructs a new node with the passed node data and an empty list of
    * children.
    *
    * @param nodeData instance of node data (node label).
    */
   public Node(D nodeData) {
-    this.children = new Vector<Node<D>>();
+    this.children = new ArrayList<>();
```
Comment (mateuszpawlik) [inline src/main/java/at/unisalzburg/apted/node/Node.java]: Shouldn't it be 'ArrayList<Node<D>>'?

**Review by mateuszpawlik — CHANGES_REQUESTED** [review]
I've commented some of the changes. Highlights:
- the packaging prefix should be 'at.unisalzburg.dbresearch',
- automatically generated files should go away from the repository (they should not be added in the first place),
- some files have only EOL removed - this should not be part of a commit,
- some other minor things.

File: .gitignore (hunk 1 above)
Comment (miho) [reply .gitignore]: true

File: build.gradle (hunk 3 above)
Comment (miho) [reply build.gradle]: Sure it will affect current users. Changing the package name is a breaking API change. On the other hand, the current situation is causing similar problems. To simplify the migration you could ship a version that contains both, the old packages and the new ones. In the old, you add a warning that will tell users that they have to migrate to the new package name.

File: gradle/wrapper/gradle-wrapper.properties (hunk 4 above)
Comment (miho) [reply gradle/wrapper/gradle-wrapper.properties]: This file belongs to the gradle wrapper. The fact that it is automatically created isn't relevant here. The sole purpose is to reduce the complexity of the build process and to prevent potential problems with incompatible gradle versions.

Repositories that contain gradle wrapper scripts and configs do not require a manual gradle download. Users can simply run `./gradlew.sh assemble` to compile the code. If the required gradle version is not already installed it will be downloaded.

This is a well accepted procedure. You can read the documentation here: https://docs.gradle.org/current/userguide/gradle_wrapper.html

File: gradlew (hunk 5 above)
Comment (miho) [reply gradlew]: see above

File: src/main/java/at/unisalzburg/apted/distance/AllPossibleMappingsTED.java (hunk 8 above)
Comment (miho) [reply src/main/java/at/unisalzburg/apted/distance/AllPossibleMappingsTED.java]: well spotted

File: src/main/java/at/unisalzburg/apted/node/Node.java (hunk 9 above)
Comment (miho) [reply src/main/java/at/unisalzburg/apted/node/Node.java]: No, not necessarily. Java infers this type info from `this.children`. Declaring it in both places is redundant.

File: settings.gradle (hunk 7 above)
Comment (miho) [reply settings.gradle]: This file is new. It is the proper way to name a gradle project. What do you mean by 'it has not changed'?

File: .gitignore (hunk 1 above)
Comment (mateuszpawlik) [reply .gitignore]: Sorry for that. My mistake. Thanks for the explanation.

File: gradle/wrapper/gradle-wrapper.properties (hunk 4 above)
Comment (mateuszpawlik) [reply gradle/wrapper/gradle-wrapper.properties]: Thank you for your expertise. I've read the documentation you linked to. I'm not Java developer but I don't agree with you. Automatically generated files don't have place in my repository. I've looked around and found this constructive discussion: https://stackoverflow.com/questions/20348451/why-should-the-gradle-wrapper-be-committed-to-vcs

Please read through it and propose a solution that suits you. I personally like the answer by Tomas Bjerre. I'm sure we can find a solution.

File: settings.gradle (hunk 7 above)
Comment (mateuszpawlik) [reply settings.gradle]: Again my mistake. The documentation has changed heavily since I read it. This file is fine.

File: gradle/wrapper/gradle-wrapper.properties (hunk 4 above)
Comment (miho) [reply gradle/wrapper/gradle-wrapper.properties]: From a theoretical point of view I understand your aversion against automatically generated files. But I think you are giving the automatic vs. non-automatic discussion way too much attention. In my opinion it should be a pragmatic discussion about pros vs. cons. The discussion you referenced is more about opinions than pragmatic solutions.

In particular, the comparison of the JDK (a **platform dependent** dependency that weights roughly **1GB** if you want your library to work on Linux, Mac and Windows, only counting x64 arch) and a **platform independent** dependency that is roughly **60-70KB** is very strange and misleading. As a pragmatic developer I always look at pros vs. cons. The cons of a nearly 1GB binary blob outweighs by far the benefit it could potentially have. A 70KB platform independent way to solve my dependency problems has certainly more pros than cons.

In his reply, Thomas asks

> "Do you also commit all your dependencie libraries?"

This is where he completely misses the intention behind the Gradle wrapper. It is not some random dependency. It bootstraps your build process. All of your other dependencies (including Gradle itself) are automatically downloaded. It's all about automation at the cost of just 70KB.

Installing Gradle manually and then calling the wrapper is possible. I did this for years and I had a rather negative experience with that approach. It is likely to introduce problems that haven't been addressed in the discussion. Not long ago, I got tons of questions related to [Could not determine java version from ‘9.0.1’](https://discuss.gradle.org/t/could-not-determine-java-version-from-9-0-1/24457) and similar problems. I learned my lesson ;)

Anyway, if you are still not convinced, just throw those files away and rebase the things you like into master. For me this is not a big deal. I will keep the wrapper in our fork since it works great for us.

**Comment by mateuszpawlik** [comment]
Thanks for all the points. To finalise quickly, I've decided to make the changes myself. Here's the summary of what I'm planning to do.
(1) I'll apply your changes to the Gradle part. However, I will not push the automatically generated files. For that I can offer to publish a tarball if it helps. I'll update Gradle version to the newest. I'll update the dependencies versions. 
(2) I'll change the packaging to avoid conflicts and add the prefix 'at.unisalzburg.dbresearch'. I'll leave the entry points of the old API and annotate as deprecated. That should simplify the transition for everyone using it.
(3) I'll update README and javadoc respectively.
Does that sound reasonably?

**Comment by miho** [comment]
(1) I said above this is fine for us. We will provide the files in our fork. So there's no need for an additional archive. Such an archive would have to track changes between versions, e.g. Gradle 4 -> 5 etc. which is hard to achieve and makes thigs even more complex.
(2) I would recommend to merge the Vector->ArrayList & Generic fixes as well.
(3) Works for me.

**Comment by mateuszpawlik** [comment]
I've pushed a branch `project-wide-change` with all the changes we've discussed. As I've mentioned, it was easier for me this way due to the nature of changes. I'll merge with your approval.

**Comment by miho** [comment]
The changes look good to me. We did some early tests with you algorithm and so far it works great.

**Comment by mateuszpawlik** [comment]
I've pushed my changes to master. Can we close this pull request?

@miho Another question. Could you write me what exactly do you use our algorithm for?

**Comment by miho** [comment]
Thanks for adding the proposed changes to the master branch.
